classes/i18n/Locale.inc.php

```diff
@@ -0,0 +1,428 @@
+<?php
+declare(strict_types = 1);
+
+/**
+ * @defgroup i18n I18N
+ * Implements localization concerns such as locale files, time zones, and country lists.
+ */
+
+/**
+ * @file classes/i18n/Locale.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class Locale
+ * @ingroup i18n
+ *
+ * @brief Provides methods for loading locale data and translating strings identified by unique keys
+ */
+
+namespace PKP\i18n;
+
+use Illuminate\Support\Facades\App;
+use PKP\config\Config;
+use PKP\core\Registry;
+use PKP\core\PKPRequest;
+use PKP\db\DAORegistry;
+use PKP\i18n\interfaces\LocaleInterface;
+use PKP\plugins\HookRegistry;
+use PKP\plugins\PluginRegistry;
+use PKP\session\SessionManager;
+use Illuminate\Support\Facades\Cache;
+use PKP\i18n\translation\LocaleBundle;
+use DateTime;
+use DomainException;
+use InvalidArgumentException;
+use LogicException;
+use SplFileInfo;
+use RecursiveDirectoryIterator;
+use RecursiveIteratorIterator;
+use RecursiveRegexIterator;
+use RegexIterator;
+use SimpleXMLElement;
+
+class Locale implements LocaleInterface
+{
+    /** @var string Keeps the locales available in the system */
+    protected const LOCALE_REGISTRY_FILE = 'registry/locales.xml';
```

Comment:
I'd like to remove this XML file and add the locales through code, is there something against?

Reply:
Can you describe what you have in mind? See related: https://github.com/pkp/pkp-lib/issues/6080

Reply:
Given that this file is used only by the `getLocales()` method.

- I was going to build the list of available locales based on the existing locale files.
- And about the "metadata", I thought about using a package, but I didn't find something that address all the needs in a fast search, just:
  - https://github.com/leodido/langcode-conv: has iso639-1, iso639-2/t, iso639-2/b and iso639-3, but not the direction (rtl/ltr)
  - https://punic.github.io: has a lot of locale information, inclusive the direction, but not the other representations

If there isn't a suitable package, I'd just transform the file into code:
```php
return [
  new LocaleMetadata(...),
  ...
];
```

---

About the issue, it's possible to replace the `Locale` in the Laravel container and provide a customized implementation.

Reply:
We already have ISO639-3 included in OJS/OMP/OPS as part of https://github.com/sokil/php-isocodes (which is already used for countries and currencies).

I think the features that'll be hard to keep are:
- LRT vs RTL
- Completeness indication (I suppose this could be gleaned when generating the cache; currently we go by a heuristic where around 90% of the strings in the English language must be covered by the translation to be considered complete)
- Country-specific stylesheet (this can be optionally specified with the `stylesheet` attribute, and is useful for RTL languages)

Reply:
- LTR/RTL: I'll check if there's a small package providing it, I saw that "punic" package has, but I didn't test.
- Completeness: Are we using this somewhere? I think it would be useful to have it displayed on the locale selection (90% complete. [Contribute with translations here])
- Country-specific stylesheet: I removed this just to create a discussion. Do you have any idea about the usage? I think it should be implemented as a file upload in the administrative page 🤔

Reply:
Completeness is currently displayed on the installation form and will let users know what to expect, including helping them choose between competing variants (e.g. `pt_PT` and `pt_BR`; `fr_CA` and `fr_FR`).

About the stylesheet: It's a language-specific stylesheet, not a country-specific one. This can be very useful with non-Latin scripts and layouts, e.g. Hebrew, Arabic, etc.

Reply:
@asmecher Anything against moving the locale stylesheet into the administration page (file upload/textarea)?
If that's ok, I can create an issue `remove the locales.xml` to address this and the other items.

Reply:
I think the idea was that we would get translators/user communities to contribute stylesheets that would be shipped as part of the application -- e.g. an `ar_IQ`-specific stylesheet that fixes display issues for that language. That hasn't happened yet, but I think it's an important feature.

Reply:
Great =]
@asmecher did you see the other comments that I left in this PR? If you've already checked them and don't have any specific observation/opinion, then just let me know... I'll be working on them next days.